### DISCUSSION TITLE
Initial Project Roadmap

### DISCUSSION THREAD

**Discussion by claymcleod** [stjude-rust-labs/sprocket#44] [Announcements]
## Current focus areas (updated 2026-04-08)

@a-frantz is working on **configuration**. This means building a new configuration resolution system from scratch, consolidating the existing configuration surface into a unified model, expanding it to cover all linting and formatting options, and documenting everything on sprocket.bio.

@Serial-ATA is working on **lint rules**. This means implementing the remaining ~9 lint rule issues (see [discussion #707](https://github.com/stjude-rust-labs/sprocket/discussions/707)) that are needed to support the stricter standards required by `openwdl/modules`.

@claymcleod is working on **WDL modules**. This means driving the [Module Specification and Ecosystem RFC](https://github.com/openwdl/wdl/discussions/700) at the spec level and building out the corresponding module resolution and import machinery in Sprocket.

@peterhuene is working on **LSP & developer experience**. This means hardening the LSP for production use, fixing stability issues, and improving the overall developer workflow across the CLI and editor integrations.

---

The following are a (semi-)prioritized list of project goals that we've set out to achieve with Sprocket. Keep in mind that not all work will occur in this repository—some will occur in [`sprocket`](https://github.com/stjude-rust-labs/sprocket) (in the `crates/` directory), some will occur in [`sprocket-vscode`](https://github.com/stjude-rust-labs/sprocket-vscode), and some will occur in new purpose-built repositories.

## Core facilities

- [x] Build a full lexer and parser (`wdl-grammar`) that is targeted at the WDL v1.2 specification.
- [x] Build a concrete syntax tree (`wdl-grammar`) implementation that will support incremental updates via red/green trees.
- [x] Build an abstract syntax tree (`wdl-ast`) to wrap all of the semantically relevant bits.
- [x] Build infrastructure for linting and validation errors (across `wdl-lint` and `wdl-ast`).
- [x] Build an analysis engine (`wdl-analysis`) that resolves cross-file dependencies.
- [x] Build an initial formatting engine (`wdl-format`).
- [x] Build an LSP that makes these lints and validation errors available in editors (`wdl-lsp`).
- [x] Write an execution engine that fully evaluates and executes WDL v1.2 workflows locally (`wdl-engine`).
- [x] Extend the execution engine to work on multiple platforms (via `wdl-engine` and `crankshaft`).
  - [x] Local execution (not to spec).
  - [x] Local Docker execution. 
  - [x] Cloud execution using the Task Execution Schema API interface.
  - [x] High-performance compute cluster execution ~~using the generic backend~~ using bespoke LSF and SLURM backends.
- [x] Create an off-the-shelf documentation tool for pleasantly releasing documentation for any WDL workflow (`wdl-doc`).

## Configuration

- [ ] Build a new configuration resolution system from scratch to unify and solidify configuration across all Sprocket tools (`sprocket.toml`).
- [ ] Provide configuration for the suite of tools above available at canonical system paths.
  - [x] Introduce configuration for enabling and disabling individual lint rules. 
  - [ ] Introduce configuration for enabling and disabling groups of lint rules. 
  - [ ] Introduce pedantic mode configuration for lint rules (global, per-tag-group, or per-rule override) — per [discussion #707](https://github.com/stjude-rust-labs/sprocket/discussions/707).
  - [x] Introduce configuration for format parameters.
  - [ ] Introduce configuration for additional format options (newline style [#667](https://github.com/stjude-rust-labs/sprocket/issues/667), call input formatting [#666](https://github.com/stjude-rust-labs/sprocket/issues/666)).
  - [x] Introduce configuration for linting parameters.
  - [x] Introduce configuration for the execution engine.
- [ ] Fully document all configuration values on sprocket.bio.

## Lint rules

New and enhanced rules needed to support the stricter standards required by `openwdl/modules` — see [discussion #707](https://github.com/stjude-rust-labs/sprocket/discussions/707).

- [x] `ModernInterpolation` — require `~{}` syntax instead of `${}` ([#710](https://github.com/stjude-rust-labs/sprocket/issues/710)).
- [x] `DenyGlobStar` — flag `glob("*")` in output blocks ([#714](https://github.com/stjude-rust-labs/sprocket/issues/714)).
- [ ] `CommandShellSafety` — require `set -euo pipefail` in command blocks ([#708](https://github.com/stjude-rust-labs/sprocket/issues/708)).
- [ ] `NoInlineInstall` — flag package manager installs in command blocks ([#709](https://github.com/stjude-rust-labs/sprocket/issues/709)).
- [ ] `ContainerUri` enhancement — upgrade mutable tag diagnostic in pedantic mode ([#711](https://github.com/stjude-rust-labs/sprocket/issues/711)).
- [ ] `ExpectedRuntimeKeys` enhancement — upgrade missing key diagnostics in pedantic mode ([#712](https://github.com/stjude-rust-labs/sprocket/issues/712)).
- [ ] `ParameterizedResources` — flag hardcoded resource values ([#713](https://github.com/stjude-rust-labs/sprocket/issues/713)).
- [ ] `OptionalInputSafety` — flag unguarded optional inputs in commands ([#715](https://github.com/stjude-rust-labs/sprocket/issues/715)).
- [ ] `NonEmptyOutputs` — flag tasks with no output declarations ([#716](https://github.com/stjude-rust-labs/sprocket/issues/716)).
- [ ] `PreferStructOverMap` — flag `Map[String, *]` where a Struct would be clearer ([#717](https://github.com/stjude-rust-labs/sprocket/issues/717)).
- [ ] `NoHostPathLiterals` — flag absolute path literals in `File`/`Directory` declarations ([#718](https://github.com/stjude-rust-labs/sprocket/issues/718)).
- [ ] Test lint rule examples in `wdl-lint`/`wdl-analysis` ([#771](https://github.com/stjude-rust-labs/sprocket/issues/771)).

## Editor support

- [x] Create a first-class extension for WDL in VSCode ([`sprocket-vscode`](https://github.com/stjude-rust-labs/sprocket-vscode)).
  - [x] Syntax highlighting to support WDL v1.2 via the TextMate grammars.
  - [x] Go to definition.
  - [x] Find all references.
  - [x] Autocompletion of struct members and variables.
  - [x] Semantic code highlighting (integrated with the LSP).
  - [x] Hover items (hover and tell the type).
  - [x] Integrate code formatting into the extension.
  - [x] Rename symbol.
  - [x] Workspace and document symbols.
  - [x] Code snippets for common WDL constructs.
  - [ ] Consider automated documentation block generation.
  - [ ] Performance and reliability improvements as needed.
- [x] Create a WDL extension for Zed ([`sprocket-zed`](https://github.com/stjude-rust-labs/sprocket-zed)).
  - [x] Syntax highlighting via [`tree-sitter-wdl`](https://github.com/stjude-rust-labs/tree-sitter-wdl).
  - [x] Go to definition.
  - [x] Find all references.
  - [x] Autocompletion of struct members and variables.
  - [x] Semantic code highlighting (integrated with the LSP).
  - [x] Hover items (hover and tell the type).
  - [x] Integrate code formatting into the extension.
  - [x] Rename symbol.
  - [x] Workspace and document symbols.
  - [ ] Code snippets for common WDL constructs.
  - [ ] Consider automated documentation block generation.
  - [ ] Performance and reliability improvements as needed.
- [x] Create a WDL plugin for Neovim ([`sprocket.nvim`](https://github.com/stjude-rust-labs/sprocket.nvim)).
  - [x] Syntax highlighting via Vim regex patterns.
  - [x] Go to definition.
  - [x] Find all references.
  - [x] Autocompletion of struct members and variables.
  - [x] Semantic code highlighting (integrated with the LSP).
  - [x] Hover items (hover and tell the type).
  - [x] Integrate code formatting into the extension.
  - [x] Rename symbol.
  - [x] Workspace and document symbols.
  - [ ] Code snippets for common WDL constructs.
  - [ ] Consider automated documentation block generation.
  - [ ] Performance and reliability improvements as needed.

## Running Workflows

- [ ] Build a cloud-based TES execution system for scalable analysis in the cloud ([`planetary`](https://github.com/stjude-rust-labs/planetary)).
- [x] Implement call caching.
- [ ] Consider dynamically generating help messages for WDL workflow/task inputs when `sprocket run` is used and adding those to the help message dynamically.
- [x] A TUI-based workflow monitoring dashboard that retrieves job status from Crankshaft ([`crankshaft-console`](https://github.com/stjude-rust-labs/crankshaft/tree/main/crankshaft-console)).
  - Initial thought: piecemeal events that the client keeps up to date or just sync the overall state of the Crankshaft daemon regularly with the client.
    - If the first one, refactor code around task state within Crankshaft (come up with events, create event enum, modify `run` method in interface, implement the events for all backends).
  - Storing the internal state if a feature is enabled in Crankshaft, working on the logic to send the current state at connection time and then regular updates (either events or full state syncs).
  - TUI using gRPC (using `tokio-console` as an example).

## Developer Experience

- [x] Introduce `sprocket dev test` for running unit tests for a WDL workspace.
- [x] Introduce `sprocket dev server` for running the HTTP API server for run execution.
- [ ] Ease the issue of Docker container image building being decoupled from pipeline testing. This probably means adding some configuration to `Sprocket.toml` that allows you to specify the location of certain Docker images to discrete Dockerfiles in your project and transparently building then before each run.
- [x] Introduce a local mode that bypasses Docker containers and uses your underlying environment during testing. This allows you to focus on iterating on the functionality by tweaking your local environment and then coming back to build/productionize a Docker image later (once you've encountered and fixed all the stumbling blocks you run into in your local environment).
- [x] Consider how we can make Sprocket the go-to tool for end-to-end testing of bioinformatics tool development. For example, if you write a WDL wrapper for your a bioinformatics tool you wrote that lives alongside that tool in the repo, you should be able to make a few end-to-end tests in the `tests/` directory, run `sprocket test`, and ensure all the results still look good (or similar). This will also encourage folks to write WDL wrappers for their tools that live in the tool repos themselves.
- [ ] Fix LSP panic when switching branches ([#680](https://github.com/stjude-rust-labs/sprocket/issues/680)).
- [ ] Better test selection for `test` CLI ([#791](https://github.com/stjude-rust-labs/sprocket/issues/791)).
- [ ] Fix `test` overwhelming the Docker daemon ([#781](https://github.com/stjude-rust-labs/sprocket/issues/781)).
- [ ] Fix panic from `test --parallelism 0` ([#797](https://github.com/stjude-rust-labs/sprocket/issues/797)).
- [ ] Fix `dev test` silently accepting bad paths ([#742](https://github.com/stjude-rust-labs/sprocket/issues/742)).

## Extensions to WDL

The following are extensions to WDL that are needed to really improve bioinformatics software development using WDL.

- [ ] Documentation should be introduced using comments. Deprecate the input and output parameter information that is part of the language.
- [ ] Support for symbolic importing should be introduced such that you can import libraries from a package manager (without needing a HTTP or local path), effectively allowing the decoupling dependency management from the language — see the [Module Specification and Ecosystem RFC](https://github.com/openwdl/wdl/discussions/700).

**Comment by Serial-ATA** [comment]
@claymcleod Do you know if anyone else was interested in `sprocket-intellij`? If so that could be added to the list.
